repairDuplicateVersions.js

```diff
@@ -0,0 +1,447 @@
+                error: err.message,
+            });
+            return cbOnce(errors.InternalError);
+        });
+        return undefined;
+    });
+    sourceReq.once('error', err => {
+        log.error('error sending request to sproxyd', {
+            objectUrl,
+            sproxydKey,
+            error: err.message,
+        });
+        return cbOnce(errors.InternalError);
+    });
+    sourceReq.end();
+}
+
+function repairObject(objInfo, cb) {
+    async.mapValues({
+        objectUrl: objInfo.objectUrl,
+        objectUrl2: objInfo.objectUrl2,
+    }, (url, key, done) => {
+        fetchObjectMetadata(url, (err, md) => {
+            if (err) {
+                log.error('error fetching object location', {
+                    objectUrl: url,
+                    error: err.message,
+                });
+                return done(err);
+            }
+            if (!Array.isArray(md.location)) {
+                const msg = 'location field is not an array';
+                log.error(msg, {
+                    objectUrl: url,
+                });
+                return done(new Error(msg));
+            }
+            const locationKeys = new Set(md.location.map(loc => loc.key));
+            return done(null, { md, locationKeys });
+        });
+    }, (err, results) => {
+        if (err) {
+            return cb(err);
+        }
+        const copiedKeys = {};
+        return async.eachSeries(results.objectUrl.locationKeys, (sproxydKey, done) => {
+            if (!results.objectUrl2.locationKeys.has(sproxydKey)) {
+                // sproxyd key is not duplicated
+                return done();
+            }
+            // sproxyd key is duplicated, need to copy the data to a
+            // new key and update metadata for objectUrl
+            return copySproxydKey(objInfo.objectUrl, sproxydKey, (err, newKey) => {
+                if (err) {
+                    return done(err);
+                }
+                log.info('sproxyd key copied', {
+                    objectUrl: objInfo.objectUrl,
+                    sproxydKey,
+                    newKey,
+                });
+                copiedKeys[sproxydKey] = newKey;
+                return done();
+            });
+        }, err => {
+            if (err) {
+                return cb(err);
+            }
+            if (Object.keys(copiedKeys).length === 0) {
+                log.info('skip object already repaired', {
+                    objectUrl: objInfo.objectUrl,
+                });
+                status.objectsSkipped += 1;
+                return cb();
+            }
+            const objMD = results.objectUrl.md;
+            objMD.location.forEach(loc => {
+                if (copiedKeys[loc.key]) {
+                    // eslint-disable-next-line no-param-reassign
+                    loc.key = copiedKeys[loc.key];
+                }
+            });
+            return putObjectMetadata(objInfo.objectUrl, objMD, err => {
```

Comment:
On non-versioned buckets, could this mean that out of bad luck there could be a race condition if the same object is being actively written to, or is there a guard in bucketd/MD?

If there's a risk, maybe the doc could mention that it's preferable to shut down incoming requests while this script is running.

Reply:
This script is only relevant for versioned buckets, unless we discover a new bug that affects nonversioned buckets in the same way (unlikely). But versions could still be modified by `PutObjectAcl` or `PutObjectTagging` so the problem can exist.

Yes we could mention in the docs that it can be safer to shut down incoming traffic as the script should be quick to run it should be acceptable, even if the risk looks small since the number of affected objects should hopefully be very small compared to the total number of objects.